src/decide.proto

```diff
@@ -0,0 +1,78 @@
+syntax = "proto3";
+import "google/protobuf/empty.proto";
+
+package decide;
+
+/* The payload for a requested state change to a component. Component state may
+   be a scalar (int32) or a label (string) */
+message StateChange {
+  string component = 1;
+  oneof state {
+    int32 scalar = 4;
+    string label = 5;
+  }
+}
+
+/* The payload for a requested change to the parameters for a component */
+message ComponentParams {
+  string component = 1;
+  // The documentation for the component must define how the parameter value will be parsed from the string.
+  map<string, string> parameters = 2;
+}
+
+/* This is what gets sent on the wire */
+message Request {
+  string protocol = 1;
+  oneof request {
+    // request change to state of a component. Reply type is ok or error.
+    StateChange state_change = 2;
+    // request reset to default state; value is name of component. Reply type is ok or error
+    string state_reset = 3;
+    // request lock on experiment; value is name of experiment. Reply type is ok or error
+    string lock_expt = 16;
+    // request unlock of experiment. Reply type is ok or error
+    google.protobuf.Empty unlock_expt = 17;
+    // request shutdown of controller. Reply type is ok or error
+    google.protobuf.Empty shutdown = 18;
+    // request update to parameter values. Reply type is ok or error.
+    ComponentParams set_parameters = 19;
+    // request parameter values for component. Reply type is params or error.
+    string get_parameters = 20;
+    // request config values for component. Reply type is config or error.
+    string get_component_config = 21;
+    // request config identifier for the controller. Reply type is config or error.
+    google.protobuf.Empty get_controller_config = 22;
+  }
+}
+
+/* These are the reply types */
+message Reply {
+  string protocol = 1;
+  oneof reply {
+    // For state_change, state_reset, lock_expt, unlock_expt, shutdown:
+    // indicates the request was correctly formed and was acted on
+    google.protobuf.Empty ok = 2;
+    // indicates an error with the request, contents give the cause
+    string error = 3;
+    // reply to get_parameters
+    map<string, string> params = 19;
+    // reply to get_component_config, and get_controller_config
+    map<string, string> config = 20;
+  }
```

Comment:
I think we're missing the reply to `get_controller_config`

It might be worthwhile to write [services](https://developers.google.com/protocol-buffers/docs/proto3#services) to define each request/reply type pair. Then we could set up prost to [automatically generate stub methods](
https://docs.rs/prost-build/0.8.0/prost_build/trait.ServiceGenerator.html).

Reply:
That looks interesting. Is the implementation in prost mature enough to be worth it? Automatic code generation could be useful, but I'm guessing we still need a bunch of glue to wire it up to zmq.

Reply:
> I think we're missing the reply to get_controller_config

Controller config is just a string? If so, yes, we need that type too.

Reply:
> Is the implementation in prost mature enough to be worth it?

Not sure, I'll have to look closer. It was appealing to me as well just because it makes explicit which requests are associated with which replies

Reply:
> Controller config is just a string? If so, yes, we need that type too.


The comment above `get_controller_config` says it returns an identifier for the config, but I guess it could return contents instead

Reply:
Also, map fields aren't allowed inside of oneofs, so we'll have to figure out a different strategy if we want to keep it this way